pkg/kubectl/cmd/edit.go

```diff
@@ -0,0 +1,419 @@
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(filenames...).
+		//SelectorParam(selector).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest()
+	if err != nil {
+		return err
+	}
+
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	r := b.Flatten().Do()
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+
+	defaultVersion := cmdutil.OutputVersion(cmd, clientConfig.Version)
+	results := editResults{}
+	for {
+		obj, err := resource.AsVersionedObject(infos, false, defaultVersion)
+		if err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+
+		// TODO: add an annotating YAML printer that can print inline comments on each field,
+		//   including descriptions or validation errors
+
+		// generate the file to edit
+		buf := &bytes.Buffer{}
+		if err := results.header.WriteTo(buf); err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		if err := printer.PrintObj(obj, buf); err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+		original := buf.Bytes()
+
+		// launch the editor
+		edit := editor.NewDefaultEditor()
+		edited, file, err := edit.LaunchTempFile("oc-edit-", ext, buf)
+		if err != nil {
+			return preservedFile(err, results.file, cmd.Out())
+		}
+
+		// cleanup any file from the previous pass
+		if len(results.file) > 0 {
+			os.Remove(results.file)
+		}
+
+		glog.V(4).Infof("User edited:\n%s", string(edited))
+		lines, err := hasLines(bytes.NewBuffer(edited))
+		if err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+		if bytes.Equal(original, edited) {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, cmd.Out())
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, no changes made.")
+			return nil
+		}
+		if !lines {
+			if len(results.edit) > 0 {
+				preservedFile(nil, file, cmd.Out())
+			} else {
+				os.Remove(file)
+			}
+			fmt.Fprintln(cmd.Out(), "Edit cancelled, saved file was empty.")
+			return nil
+		}
+
+		results = editResults{
+			file: file,
+		}
+
+		// parse the edited file
+		updates, err := rmap.InfoForData(edited, "edited-file")
+		if err != nil {
+			results.header.reasons = append(results.header.reasons, editReason{
+				head: fmt.Sprintf("The edited file had a syntax error: %v", err),
+			})
+			continue
+		}
+
+		visitor := resource.NewFlattenListVisitor(updates, rmap)
+
+		// need to make sure the original namespace wasn't changed while editing
+		if err = visitor.Visit(resource.RequireNamespace(cmdNamespace)); err != nil {
+			return preservedFile(err, file, cmd.Out())
+		}
+
+		// attempt to calculate a delta for merging conflicts
+		delta, err := jsonmerge.NewDelta(original, edited)
+		if err != nil {
+			glog.V(4).Infof("Unable to calculate diff, no merge is possible: %v", err)
+			delta = nil
+		} else {
+			delta.AddPreconditions(jsonmerge.RequireKeyUnchanged("apiVersion"))
```

Comment:
Kind actually is somewhat dangerous to change, as is name, because they will not convey the users original intent (edit resource X of type Y) to the server.  The server would give a message like "Y is not a valid kind" or "X does not exist", not the more appropriate "you cannot rename X".

We should at minimum check "kind" - name may not exist on all API objects you edit.